feat: Add support for REPLs

philsc opened the pull request:
This patch adds a new target that lets users invoke a REPL for a given
`PyInfo` target.

For example, the following command will spawn a REPL for any target
that provides `PyInfo`:
```console
$ bazel run --//python/config_settings:bootstrap_impl=script //python/bin:repl --//python/bin:repl_dep=//tools:wheelmaker
Python 3.11.1 (main, Jan 16 2023, 22:41:20) [Clang 15.0.7 ] on linux
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>> import tools.wheelmaker
>>>
```

If the user wants an IPython shell instead, they can create a file like this:
```python
import IPython
IPython.start_ipython()
```
Then they can set this up in their `.bazelrc` file:
```
# Allow the REPL stub to import ipython. In this case, @my_deps is the name
# of the pip.parse() repository.
build --@rules_python//python/bin:repl_stub_dep=@my_deps//ipython
# Point the REPL at the stub created above.
build --@rules_python//python/bin:repl_stub=//path/to:ipython_stub.py
```

arrdem commented:
Broadly in favor of this, had a similar macro feature aligned with the verbs pattern at a former employer.

I don't love having magical behavior that runs ipython if it's on the path, IMO an ipython shell or any other notebook-like thing is a different kind of target than a vanilla shell and we shouldn't overload the one with try-import behavior which is generally an antipattern. IMO the kind of shell to run should be explicitly selected either by having separate targets or by having a feature flag we can `select()` on.

While REPL is the more familiar term to me from the Lisp world, the Python ecosystem appears to prefer the words "shell" or "console". My previous macros emitted `.shell` and `.ishell` verbs which I would suggest is slightly better aligned.

aignas commented:
What is the difference between 2 and 3? Could we just get away with 3?

philsc commented:
> I don't love having magical behavior that runs ipython if it's on the path, IMO an ipython shell or any other notebook-like thing is a different kind of target than a vanilla shell

I generally agree, but everywhere I've implemented this (3 places), one of the first questions is "how do I get ipython?". So I've just given up on that front and made these automatically use that. I can be convinced to drop it, but I suspect without it, folks are just going to implement it themselves which reduces the value of having it in a shared place.

> IMO the kind of shell to run should be explicitly selected either by having separate targets or by having a feature flag we can select() on.

Perhaps we could add an additional label_flag to let the user specify a .py file that invokes the actual shell (to replace `code.interact()`) in addition to the dep that pulls in ipython.

philsc commented:
> What is the difference between 2 and 3? Could we just get away with 3?

The difference is convenience. I have found that folks picked up the "bazel run this library with .repl at the end" a whole lot easier and even taught it to other folks.

philsc commented on python/private/py_library_macro.bzl:
```diff
@@ -14,8 +14,35 @@
+        name = name,
+        **(library_only_attrs | kwargs)
+    )
+    py_binary_rule(
```
WDYT about hiding it behind a `generate_repl = False` attribute? And then tagging it `manual`?

Alternatively, I can document how users can do this themselves. It is mighty convenient.

aignas replied:
I am +1 for users defining their own macros in this particular case.

philsc commented on python/private/sentinel.bzl:
```diff
@@ -29,6 +31,7 @@ def _sentinel_impl(ctx):
         SentinelInfo(),
         # Also output ToolchainInfo to allow it to be used for noop toolchains
         platform_common.ToolchainInfo(),
+        PyInfo(transitive_sources=depset()),
```
Fair enough. I will just point it at an empty `py_library` since there's no need to distinguish.

philsc commented on python/private/stage1_bootstrap_template.sh:
```diff
@@ -238,6 +238,10 @@ command=(
   "$@"
 )
 
+# Point libedit/readline at the correct terminfo databases.
+# https://github.com/astral-sh/python-build-standalone/blob/f0abfc9cb1f6a985fc5561cf5435f7f6e8a64e5b/docs/quirks.rst#backspace-key-doesnt-work-in-python-repl
+export TERMINFO_DIRS=/etc/terminfo:/lib/terminfo:/usr/share/terminfo
```
Are you okay with me doing this for just the REPL-specific targets? Otherwise, I see this as a source of frustration for users.

philsc replied:
At least in my local env, rules_python started pulling in a new enough version that it's not needed anymore.

> Starting in the first release after 20240107, the Linux distributions are configured to automatically use the terminfo database in /etc/terminfo, /lib/terminfo, and /usr/share/terminfo.

So I removed it.

philsc commented on python/private/stage2_bootstrap_template.py:
```diff
@@ -365,6 +365,16 @@ def main():
+        global MAIN_PATH
+        global MAIN_MODULE
+        MAIN_PATH = ""
+        # TODO(philsc): Can we point at python.bin.repl instead? That would mean
+        # adding it as a dependency to all binaries.
```
Do you have a thought on injecting that code into here?
I could copy-paste it. Similar to the runfiles code?

philsc commented:
> While REPL is the more familiar term to me from the Lisp world, the Python ecosystem appears to prefer the words "shell" or "console". My previous macros emitted .shell and .ishell verbs which I would suggest is slightly better aligned.

Hmm. Perhaps I'm old :stuck_out_tongue: 

REPL does appear in the docs: https://docs.python.org/3/tutorial/appendix.html#tut-interac

> There are two variants of the interactive [REPL](https://docs.python.org/3/glossary.html#term-REPL). The classic basic interpreter is supported on all platforms with minimal line control capabilities.

And there's even an environment variable for configuration:
https://docs.python.org/3/using/cmdline.html#envvar-PYTHON_BASIC_REPL

I'm happy to change it to ".shell" or similar if you all want, but I don't think REPL is un-Pythonic so to speak.

aignas commented:
One extra question about what is the difference between your `//python/bin:repl` target and the existing `//python/bin:python` target? I have a usecase where I need a handle to the python interpreter with all of the deps available and importable. The current `//python/bin:python` only gives me a clean interpreter but the deps are not be found in my `PYTHONPATH`.

Should we have the approach where we:
1. Fix the `//python/bin:python` to have the deps?
2. If not, have a `//python/bin:repl` target to start with that just unblocks us from having a working interpreter.
3. Add the `env var` to the bootstrap (which could be super useful but the API might need to be discussed further.
4. Discuss the macro documentation, etc later.

philsc commented:
> One extra question about what is the difference between your `//python/bin:repl` target and the existing `//python/bin:python` target? I have a usecase where I need a handle to the python interpreter with all of the deps available and importable. The current `//python/bin:python` only gives me a clean interpreter but the deps are not be found in my `PYTHONPATH`.

Can you elaborate on this a bit? Is the use case to do something like `subprocess.run([sys.executable, "path/to/file.py"])`?

The goal with this PR here is to create a REPL in an environment identical to what you would see when running your code in a `py_binary`. I.e. all the things that the first and second stage bootstrap script do. The three that I can think of:
1. venv creation
2. safe path handling
3. setting up runfiles env vars

You don't get that with only the interpreter, even if it could import the deps.

I think if we can figure out how to move all those things into a pre-generated venv, then that could work. But it's not obvious to me that we can do that at this time.

> 
> Should we have the approach where we:
> 
>     1. Fix the `//python/bin:python` to have the deps?
> 
>     2. If not, have a `//python/bin:repl` target to start with that just unblocks us from having a working interpreter.
> 
>     3. Add the `env var` to the bootstrap (which could be super useful but the API might need to be discussed further.
> 
>     4. Discuss the macro documentation, etc later.

Removed the env var in the bootstrap script and the macro stuff.

aignas commented:
The concrete usecase is that I am using `dagster` and `dagster` allows users to include user code deployments via specifying in a YAML directory a list of `venv` with a path to the python interpreter.

Right now I am hacking this around by having a `py_library` with all the deps and then a `py_binary` that has the `main` as the following file:
```python
import subprocess
import sys

if __name__ == "__main__":
    subprocess.run([sys.executable] + sys.argv[1:])  # noqa: S603
```

This basically re-execs the interpreter from the environment with any parameters that may be passed to the target. If I run this target I get the `REPL`. And I can do `import dagster` in the said vanila `python` REPL.

If on the other hand I pass the `py_library` to the `@rules_python//python/bin:python` target via a the label flag and try doing `import dagster` it fails because the deps are not wired in.

My question here is:
* Why do we have `@rules_python//python/bin:python` in addition to this new REPL target?
* If there is a reason to have both, where should the user use one over the other?

groodt commented:
My 2c here. I would prefer to have this in user-space. We have a similar need at work, and we added a small macro to create an ipython py_console_script_binary with the dependencies. 

This deals with most of the concerns listed above?
* venv creation ✅ - I presume you really mean the runfiles or PYTHONPATH? But broadly, when the rules support site-packages layout, you can get this for free with a user-space solution
* safe path handling ✅ - A user-space solution is launched with the same SAFEPATH handling
* setting up runfiles env vars ✅ - A user-space solution sets up runfiles

Im worried about adding *more* complexity into the toolchains and bootstrapping areas at the moment.

If your main motivation is convenience for users, it's unclear why you couldn't create some user facing macros in your repository (automated or not via gazelle) that wrap the native py_library and py_binary targets with an ipython repl and targetnames of your choosing?

aignas commented:
I am curious if having the `main_module` with a select statement something like:
```
py_binary(
    name = "my_py_binary",
    srcs = ...,
    deps = ...,
    main_module = select({
        "@rules_python//python/config_settings:launch_repl": "$(PYTHON3)",
        "//conditions:default": "",
    }),
)
```

I agree with @groodt here that running `IPython` should be solved in the user space, but launching vanilla python should be a part of the core rules IMHO, but I am not sure exactly how.

I personally liked the already existing target `@rules_python//python/bin:python` but currently it is not setting up the `PYTHONPATH` yet.

philsc commented:
> My question here is:
> 
>     * Why do we have `@rules_python//python/bin:python` in addition to this new REPL target?

The goal with this PR here is to create a REPL in an environment identical to what you would see when running your code in a py_binary. I.e. all the things that the first and second stage bootstrap script do. The three that I can think of:

1. venv creation
2. safe path handling
3. setting up runfiles env vars

You don't get that with only the interpreter, even if it could import the deps.

If we want to to make `@rules_python//python/bin:python` capable of importing arbitrary py_library targets, then we'd have to teach `python/private/interpreter_tmpl.sh` about runfiles and set up the PYTHONPATH. I suspect it would look a lot like what `python/private/python_bootstrap_template.txt` looks like right now (maybe minus the zipfile stuff? :thinking: ).  If that's what we want to do that, that's okay, but that's not the goal of this target here. The goal of the `python/bin:python` target was to just give access to the interpreter

> 
>     * If there is a reason to have both, where should the user use one over the other?

When you just want the interpreter (for deploying, bootstrapping, etc.) you use `@rules_python//python/bin:python`. For development, when you want a REPL in an environment identical to what you see in a `py_binary`, then you use `python/bin:repl`.

Your work around with a `py_binary` target sounds reasonable to me. If we turn that into a macro, then we can avoid introducing a transition for targets that want the interpreter plus some libraries.

> My 2c here. I would prefer to have this in user-space. ... Im worried about adding more complexity into the toolchains and bootstrapping areas at the moment.

I removed all the changes to IPython, the bootstrapping scripts, and there are no changes to the toolchain stuff. Could I ask you to double check that you're still interested in seeing some changes?

> If your main motivation is convenience for users, it's unclear why you couldn't create some user facing macros in your repository (automated or not via gazelle) that wrap the native py_library and py_binary targets with an ipython repl and targetnames of your choosing?

I removed all the IPython-related changes. The user can create their own macro or substitute their own stub if they like.

> I am curious if having the main_module with a select statement something like:
> ```
> py_binary(
>     name = "my_py_binary",
>     srcs = ...,
>     deps = ...,
>     main_module = select({
>         "@rules_python//python/config_settings:launch_repl": "$(PYTHON3)",
>         "//conditions:default": "",
>     }),
> )```

That's actually pretty cool! I think that could work, but it would be limited to `py_binary` targets.  It's not immediately obvious to me how to extend that to py_library without doing, say, a user-space wrapping macro that automatically adds a `py_binary` wrapper for the target (similar to what Greg was saying).

aignas commented on python/private/repl.bzl:
```diff
@@ -0,0 +1,67 @@
+    },
+)
+
+def py_repl_binary(name, stub, deps = [], data = [], **kwargs):
```
Please document the values with the usual docs.

philsc replied:
Done

aignas commented on python/private/repl.bzl:
```diff
@@ -0,0 +1,67 @@
+        "out": attr.output(
+            mandatory = True,
+        ),
+        "stub": attr.label(
+            mandatory = True,
+            allow_single_file = True,
```
What is `stub`?

aignas commented on python/private/repl.bzl:
```diff
@@ -0,0 +1,67 @@
+# Copyright 2025 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
```
The Copyright can be removed from the PR, `bazel-contrib` does not require that.

aignas commented on python/bin/repl_stub.py:
```diff
@@ -0,0 +1,3 @@
+import code
+
+code.interact()
```
What is `code` and what is `code.interact`? Is it something special? Could you please add a comment?

aignas commented on python/bin/BUILD.bazel:
```diff
@@ -22,3 +23,35 @@ label_flag(
+    build_setting_default = "repl_stub.py",
+)
+
+# The user can modify this flag to make an interpreter shell library available
+# for the stub. E.g. if they switch the stub for an ipython-based one, then they
+# can point this at their version of ipython.
```
In our `docs` we have documentation for targets, e.g. the config settings, so these comments could be written there so that the user can easily find the docs.

philsc replied:
Will do hopefully mid-next week.

aignas reviewed (commented):
As far as the REPL implementation goes, I don't have major issues with this, but love to discuss on the comments I left.

aignas commented:
Changed to `Ready for Review` since we are actively reviewing the PR.

aignas commented on python/bin/repl_stub.py:
```diff
@@ -1,3 +1,18 @@
+
+The logic for PYTHONSTARTUP is handled in python/private/repl_template.py.
+"""
+
```
Could we also copy the logic for the banner, so that users are greeted with the familiar Python version?

philsc replied:
Done.

That is actually surprisingly awkward. That code is implemented a few times. At least once in the cpython startup code and once in the `code` module.  Since there's no good way to invoke either of those from here, I did indeed end up copying it. Let me know if this is what you had in mind.

aignas replied:
Sorry, did not realize it was a little awkward, but thanks for doing it!

aignas commented on python/private/repl.bzl:
```diff
@@ -33,26 +20,54 @@ _generate_repl_main = rule(
     )
 
-    py_binary(
+    _py_binary(
```
nit: I remember @rickeylev mentioning to avoid `out` parameters because then we can use providers to pass the files around more cleverly. I think the `_generate_repl_main` could do that, but if you disagree on this approach, then it is fine to leave as is here for now.

philsc replied:
Done

It did make the filename a bit weird though. Instead of `repl.py`, it's now called `repl_py.py`. Will have to ask Richard about that. Unless I misunderstood what you had in mind, apologies.

aignas reviewed (commented):
The only question is the banner message, what do you think abut adding it?

aignas commented on python/bin/BUILD.bazel:
```diff
@@ -22,3 +23,35 @@ label_flag(
+# can point this at their version of ipython.
+label_flag(
+    name = "repl_stub_dep",
+    build_setting_default = "//python/private:empty",
```
I am fine to move forward with the current impl.

aignas reviewed (approved):
LGTM, the last thing is to add a CHANGELOG entry in the unreleased "Added" section.

Thanks for doing this!

philsc commented:
Thanks for finishing this up @aignas! Really appreciate it :heart:
